R/approx_distance.R

```diff
@@ -0,0 +1,233 @@
+      # belonging to G
+      a0 <- c(rep(1, length(q_F)), rep(-1, length(q_G)))
+      
+      # re-order both vectors:
+      q <- q0[order(q0)]
+      a <- a0[order(q0)]
+      # and compute "how many quantiles ahead" F or G is at a given segment:
+      b <- abs(cumsum(a))
+      
+      # compute the lengths of segments defined by sorted quantiles:
+      diffs_q <-
+        c(diff(q), 0) # zero necessary for indexing below, but we could put
+      # anything (gets multiplied w zero)
+      
+      # and approximate CD
+      if (approx_rule == "equal_1") {
+        cvm <- sum(diffs_q * b * (b + 1)) / (K + 1) / (K)
+      } else if (approx_rule == "equal_2") {
+        cvm <- sum(diffs_q * b ^ 2 / K ^ 2)
+      }
+      metric <- mean(cvm)
+    }
+    else if (approx_rule == "left-sided" |
+             approx_rule == "trapezoid") {
+      N <- length(q_F)
+      M <- length(q_F)
+      # pool quantiles:
+      q0 <- c(q_F, q_G)
+      a0 <- c(rep(1, N), rep(0, M))
+      # indicator whether the entry is from F or G
+      a <- a0[order(q0)]
+      q <- q0[order(q0)]
+      tau <- c(tau_F, tau_G)[order(q0)]
+      diffs_q <- diff(q)
+      # prob level vectors
+      tau_F_v <- a * tau
+      tau_F_v2 <- sapply(1:(N + M),
+                         function(x)
+                           ifelse(
+                             tau_F_v[x] != 0,
+                             tau_F_v[x],
+                             ifelse(min(which(tau_F_v != 0)) > 1 &&
+                                      x <= min(which(tau_F_v != 0)), 0,
+                                    tau_F_v[max(which(tau_F_v[1:x] !=
+                                                        0))])
+                           ))
+      tau_G_v <- abs(a - 1) * tau
+      tau_G_v2 <- sapply(1:(N + M),
+                         function(x)
+                           ifelse(
+                             tau_G_v[x] != 0,
+                             tau_G_v[x],
+                             ifelse(min(which(tau_G_v != 0)) > 1 &&
+                                      x <= min(which(tau_G_v != 0)), 0,
+                                    tau_G_v[max(which(tau_G_v[1:x] !=
+                                                        0))])
+                           ))
+      if (approx_rule == "left-sided") {
+        cvm <-
+          sum(((tau_F_v2[1:(N + M) - 1] - tau_G_v2[1:(N + M) - 1]) ^ 2) * diffs_q)
+      } else if (approx_rule == "trapezoid") {
+        cvm <-
+          sum((((tau_F_v2[1:(N + M) - 1] - tau_G_v2[1:(N + M) - 1]) ^ 2 + (tau_F_v2[2:(N +
+                                                                                         M)] - tau_G_v2[2:(N + M)]) ^ 2
+          ) / 2) * diffs_q)
+      }
+      metric <- cvm
+    }
+    return(metric)
+  }
+
+#' Approximate pairwise Cramér distance for a single target,
+#' date, location data frame
+#'
+#' @param single_frame a data frame containing correctly-formatted forecasts
```

Comment:
I suggest calling this something more descriptive. how about `forecast_data`? and indicating in the documentation that this is "a data frame containing forecasts formatted as returned from load_forecasts()` so you can link to the specification there?